File: features/content/affiliate/src/main/java/com/tokopedia/affiliate/feature/tracking/domain/interactor/GetByMeUseCase.java
package com.tokopedia.affiliate.feature.tracking.domain.interactor;

import android.content.Context;

import com.tokopedia.abstraction.common.di.qualifier.ApplicationContext;
import com.tokopedia.abstraction.common.utils.GraphqlHelper;
import com.tokopedia.affiliate.R;
import com.tokopedia.affiliate.feature.tracking.domain.model.Data;
import com.tokopedia.graphql.data.GraphqlClient;
import com.tokopedia.graphql.data.model.GraphqlRequest;
import com.tokopedia.graphql.data.model.GraphqlResponse;
import com.tokopedia.graphql.domain.GraphqlUseCase;

import java.util.HashMap;
import java.util.Map;

import javax.inject.Inject;

import rx.Subscriber;

public class GetByMeUseCase {

    private static final String AFFILIATE_NAME = "name";
    private static final String URL_KEY = "key";
    private GraphqlUseCase graphqlUseCase;
    private Context context;

    @Inject
    GetByMeUseCase(@ApplicationContext Context context, GraphqlUseCase graphqlUseCase) {
        GraphqlClient.init(context);
        this.context = context;
        this.graphqlUseCase = graphqlUseCase;
    }

    public void createObservable(String affiliateName, String urlKey,
                                 Subscriber<GraphqlResponse> subscriber) {
        Map<String, Object> variables = new HashMap<>();

        variables.put(AFFILIATE_NAME, affiliateName);
        variables.put(URL_KEY, urlKey);

        GraphqlRequest graphqlRequest = new GraphqlRequest(
                GraphqlHelper.loadRawString(context.getResources(), R.raw.query_url_tracking),
                Data.class,
                variables,
                false);

        graphqlUseCase.clearRequest();
        graphqlUseCase.addRequest(graphqlRequest);
        graphqlUseCase.execute(subscriber);
    }

    public void unsubscribe() {
        if (graphqlUseCase != null) {
            graphqlUseCase.unsubscribe();
        }
    }
}

File: features/content/affiliate/src/main/java/com/tokopedia/affiliate/feature/tracking/domain/interactor/GetByMeUseCase.kt
package com.tokopedia.affiliate.feature.tracking.domain.interactor

import com.tokopedia.affiliate.feature.tracking.domain.model.Data
import com.tokopedia.graphql.data.GraphqlClient
import com.tokopedia.graphql.data.model.GraphqlRequest
import com.tokopedia.graphql.data.model.GraphqlResponse
import com.tokopedia.graphql.domain.GraphqlUseCase
import rx.Subscriber
import java.util.*
import javax.inject.Inject

class GetByMeUseCase @Inject internal constructor(private val graphqlUseCase: GraphqlUseCase) {

    private val query = """
        query getRealURL(${'$'}name: String!, ${'$'}key: String!)
        {
          bymeGetRealURL(affiliateName: ${'$'}name, urlKey: ${'$'}key) {
            realURL
            error {
              type
              code
            }
          }
        }
    """

    fun createObservable(affiliateName: String, urlKey: String,
                         subscriber: Subscriber<GraphqlResponse?>) {
        val variables: MutableMap<String, Any> = HashMap()
        variables[AFFILIATE_NAME] = affiliateName
        variables[URL_KEY] = urlKey
        val graphqlRequest = GraphqlRequest(
                query,
                Data::class.java,
                variables,
                false)
        graphqlUseCase.clearRequest()
        graphqlUseCase.addRequest(graphqlRequest)
        graphqlUseCase.execute(subscriber)
    }

    fun unsubscribe() {
        graphqlUseCase.unsubscribe()
    }

    companion object {
        private const val AFFILIATE_NAME = "name"
        private const val URL_KEY = "key"
    }
}